File: src/app/features/automation/services/automation.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from '@services/api.service';
import {
  QuickAction,
  ActionStep,
  AutomationRecipe,
  ExecutionHistoryEntry,
} from '@entities/automation.model';

export type { QuickAction, ActionStep, AutomationRecipe, ExecutionHistoryEntry };

@Injectable({
  providedIn: 'root',
})
export class AutomationService {
  private api = inject(ApiService);

  constructor() {}

  async getQuickActions(): Promise<QuickAction[]> {
    try {
      const result = await this.api.invoke<QuickAction[]>('get_quick_actions');
      return result;
    } catch (error) {
      throw error;
    }
  }

  async executeAction(actionId: string): Promise<string> {
    try {
      const result = await this.api.invoke<string>('execute_action', { actionId });
      return result;
    } catch (error) {
      throw error;
    }
  }

  async getRecipes(): Promise<AutomationRecipe[]> {
    try {
      const result = await this.api.invoke<AutomationRecipe[]>('get_recipes');
      return result;
    } catch (error) {
      throw error;
    }
  }

  async saveRecipe(recipe: AutomationRecipe): Promise<string> {
    try {
      const result = await this.api.invoke<string>('save_recipe', { recipe });
      return result;
    } catch (error) {
      throw error;
    }
  }

  async deleteRecipe(recipeId: string): Promise<string> {
    try {
      const result = await this.api.invoke<string>('delete_recipe', { recipeId });
      return result;
    } catch (error) {
      throw error;
    }
  }

  async executeRecipe(recipeId: string): Promise<string> {
    try {
      const result = await this.api.invoke<string>('execute_recipe', { recipeId });
      return result;
    } catch (error) {
      throw error;
    }
  }

  async getExecutionHistory(): Promise<ExecutionHistoryEntry[]> {
    try {
      const result = await this.api.invoke<ExecutionHistoryEntry[]>('get_execution_history');
      return result;
    } catch (error) {
      throw error;
    }
  }
}

File: src/application/automation_service.rs
//! AutomationService - orchestrates automation recipes

use crate::domain::{AutomationRecipe, ExecutionHistory};

pub trait AutomationServiceTrait {
    fn get_recipes(&self) -> Result<Vec<AutomationRecipe>, String>;
    fn get_recipe(&self, id: &str) -> Result<Option<AutomationRecipe>, String>;
    fn create_recipe(&mut self, recipe: AutomationRecipe) -> Result<AutomationRecipe, String>;
    fn update_recipe(
        &mut self,
        id: &str,
        recipe: AutomationRecipe,
    ) -> Result<AutomationRecipe, String>;
    fn delete_recipe(&mut self, id: &str) -> Result<(), String>;
    fn execute_recipe(&mut self, id: &str) -> Result<ExecutionHistory, String>;
}
